[Calyx] Add Static Groups and Control

andrewb1999 opened the pull request:
Adds Calyx static groups and control as proposed [here](https://github.com/cucapra/calyx/issues/1344). Static groups and control provide a first-class way to ensure the native compiler has enough information to produce efficient, static FSMs.

The first use for these static ops within CIRCT will be improving the LoopScheduleToCalyx pass, as we can now rely on static behavior to generate efficient loop FSMs.

cgyurgyik commented on lib/Dialect/Calyx/CalyxOps.cpp:
```diff
@@ -1924,22 +2110,33 @@ LogicalResult IfOp::verify() {
+/// is present), returns None.
+template <typename OpTy>
+static std::optional<EnableOp> getLastEnableOp(OpTy parent) {
+  static_assert(std::is_same<SeqOp, OpTy>() ||
```
Suggestion:

```
namespace {

// A struct to enforce that the LHS template is one of the RHS templates.
// For example:
//   std::is_any<uint32_t, uint16_t, float, int32_t>::value is false.
template <class T, class... Ts>
struct is_any : std::disjunction<std::is_same<T, Ts>...> {};

}
```

With C++17, this should work. Then you can do:

```
is_any<OpTy, SeqOp, StaticSeqOp>::value
```

This should simplify a lot of these logical ORs.

cgyurgyik commented on test/Dialect/Calyx/emit-static.mlir:
```diff
@@ -0,0 +1,188 @@
+// RUN: circt-translate --export-calyx --split-input-file --verify-diagnostics %s | FileCheck %s --strict-whitespace
```
Can you try round-tripping with the native compiler?

https://docs.calyxir.org/fud/circt.html#native-representation-to-mlir

andrewb1999 replied:
I don't think I can do this in the CIRCT CI currently. As far as I know the Calyx native compiler is not included as an optional dependency of CIRCT. I think there's a decent argument that it should be, but that's beyond the scope of this PR.

cgyurgyik replied:
This can't be done because we don't live in the world of a monorepo. 

So, unfortunately, we need to manually test this. I'm asking if you can ensure that your new changes round trip with the native compiler, i.e., does the native compiler emit correctly emit your new dialect additions. If it is not the case, then please open a bug in the native Calyx GitHub.

andrewb1999 replied:
Oh ok, I understand what you mean now. I will test this manually.

cgyurgyik replied:
Any follow up on this?

andrewb1999 replied:
It doesn't seem to work directly. I will open an issue on the native Calyx repo.

cgyurgyik reviewed (approved):
Overall LGTM. 

From an engineering perspective, I "think" we could probably reduce code duplication by having interfaces IfOpInterface, SeqOpInterface, ParOpInterface. Then have `StaticSeqOp` and `DynamicSeqOp` (or whatever). WDYT? Maybe you already tried this?


Edit: also thanks for this work! Very cool :-)

mikeurbach commented on include/circt/Dialect/Calyx/CalyxControl.td:
```diff
@@ -258,3 +394,31 @@ def WhileOp : CalyxContainer<"while", [
+  let arguments = (ins
+    I32Attr:$count
+  );
+  let description = [{
```
Please add a description for this one, seems new and would be good to document. It's exciting to see though.

mikeurbach commented on include/circt/Dialect/Calyx/CalyxStructure.td:
```diff
@@ -547,3 +618,42 @@ def GroupGoOp : CalyxGroupPort<"group_go", [
+  HasParent<"StaticGroupOp">,
+]> {
+  let summary = "Calyx Static Cycle Op";
+  let description = [{
```
Please add a description for this one as well.

mikeurbach commented on include/circt/Dialect/Calyx/CalyxStructure.td:
```diff
@@ -547,3 +618,42 @@ def GroupGoOp : CalyxGroupPort<"group_go", [
+  }];
+  let arguments = (ins
+    I32Attr:$start,
+    OptionalAttr<I32Attr>:$end
```
For repeat and cycle, are 32 bits enough? I guess probably...

andrewb1999 replied:
Yeah I made a guess here that 32 bits would be enough. I guess it's not a big deal to switch to 64 bits, but I'm not sure if it matters for any reasonable design.

mikeurbach commented on lib/Dialect/Calyx/CalyxOps.cpp:
```diff
@@ -134,12 +135,26 @@ PortInfo calyx::getPortInfo(BlockArgument arg) {
+static bool isStaticControl(Operation *op) {
+  if (isa<EnableOp>(op)) {
+    // for enables, we need to check whether its corresponding group is static
+    auto component = op->template getParentOfType<ComponentOp>();
```
Is this `template` needed? I believe weird things have happened on windows with this.

mikeurbach commented on lib/Dialect/Calyx/Export/CalyxEmitter.cpp:
```diff
@@ -22,6 +22,8 @@
 #include "llvm/ADT/SmallSet.h"
 #include "llvm/ADT/TypeSwitch.h"
+#include "llvm/Support/FormatVariadic.h"
+#include <string>
```
This include is probably not needed if it's included transitively already.

mikeurbach reviewed (commented):
I took a quick look, and left some minor comments. Also agree with @cgyurgyik's suggestions.

Thanks for adding this, I'm excited to see these representations in CIRCT!

andrewb1999 commented:
Ok I added an interface for the If ops and addressed the other comments. I'm not sure that interfaces make sense for Seq and Par ops currently because most of the shared code is already handled by the ControlLike interface. I'm pretty sure that with the current Seq and Par ops adding an interface would introduce more code than it removes, but the interface could still make sense in the future if our uses of Seq and Par change a bit. Let me know what you think about those interfaces, but I think it's not really useful to have them now.

mikeurbach commented:
Sorry i haven't had a chance to take a second look, I will come back and review this.

andrewb1999 commented:
Does anyone have a chance to review this further soon? I think I have addressed the issues and not having this merged is starting to block me.

cgyurgyik commented on lib/Dialect/Calyx/CalyxOps.cpp:
```diff
@@ -325,6 +359,23 @@ static void eraseControlWithGroupAndConditional(OpTy op,
+/// have no more uses, they will be erased.
+template <typename OpTy>
+static void eraseControlWithConditional(OpTy op, PatternRewriter &rewriter) {
+  static_assert(std::is_same<OpTy, StaticIfOp>(),
```
I'm surprised this isn't either std::is_same_v or suffixed by ::value()

andrewb1999 replied:
I'm pretty sure is_same will automatically cast to a bool without needing to use is_same_v, but let me know if I'm wrong.

cgyurgyik replied:
https://en.cppreference.com/w/cpp/types/is_same Since C++14, you're right!

cgyurgyik commented on lib/Dialect/Calyx/CalyxOps.cpp:
```diff
@@ -850,6 +901,29 @@ void SeqOp::getCanonicalizationPatterns(RewritePatternSet &patterns,
+  // StaticSeqOp should only have static control in it
+  auto &ops = (*this).getBodyBlock()->getOperations();
+  for (Operation &op : ops) {
+    if (!isStaticControl(&op)) {
```
Maybe: If (!llvm::all_of(...

cgyurgyik reviewed (approved):
Just a question about round tripping. In general looks good.

andrewb1999 commented:
I'm going to go ahead and merge this now, but if anyone has any extra comments in the future I will be happy to address them in another PR.
